### PR TITLE
Add new soda-proxy which helps to communicate between side-cars and o…

### LINKED ISSUE #377: New Soda-CSI Plug-N-Play (soda-csi-plugin)
### Feature Improvement
Users should be able to use the third party CSI drivers using the SODA-CSI driver and experience all the feature sets of SODA.

### PR DESCRIPTION
**Pull request by asifdxtreme**
…sdsapiserver

**What type of PR is this?**
/kind new feature

**What this PR does / why we need it**:
In General side-cars needs to communicate with different api's of sodafoundation components, these side-cars are written as seprate entity and is unable to use the clients provided by sodafoundation/api to connect with apiserver.
These side-cars also has some restriction to alter their deployment and mount the cert files in the containers as the deployment is managed by vendor operators.
So to mitigate this challenge we have a `soda-proxy` which exposes the required api's needed by theses side-cars, these side-cars doesn't need auth to connect to these api's, where as `soda-proxy` internally uses the auth mechanism to connect the osdsapiserver.
By this we can enable a smooth communication between side-cars and soda components.

Currenly this PR just exposes profiles api but going forward this can exposes more api's which will be needed by these side-cars to bring the intelligence of sodafoundation to these side-cars.

**Which issue(s) this PR fixes**:

Fixes #377 

**Test Report Added?**:
 /kind TESTED

**Test Report**:
Follow the readme to bring up the soda-proxy and then run the [example](https://github.com/sodafoundation/examples/tree/master/soda-csi-plug-n-play-poc).

### CODE REVIEW

File: csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	"net/http"
+	"os"
+)
+
+func GetProfile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	key := vars["id"]
+	opensdsEndpoint, errr := os.LookupEnv("OPENSDS_ENDPOINT")
+	if !errr {
+		fmt.Println("No env variables found for endpoint, switching to default")
```
Comment (kumarashit) [inline csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go:17]: Why not use logger?

File: csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+		opensdsEndpoint = "http://127.0.0.1:50040"
+	}
+	client, err := opensds.GetClient(opensdsEndpoint, "keystone")
+	if client == nil || err != nil {
+		fmt.Printf("get opensds client failed: %v", err)
+	}
+	profile, errosds := client.GetProfile(key)
+	if errosds != nil {
+		fmt.Printf("Got error in GetProfile  : %s ===== %s", errosds.Error())
```
Comment (kumarashit) [inline csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go:26]: Please make error msg to start with lowercase

File: csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go (hunk 1 above)
Comment (asifdxtreme) [reply csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go:17]: Default logger will not work in this as the sodafoundation/api/client makes their own log-dir which creates a panic at runtime, will take this up in further iteration.

File: csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go (hunk 2 above)
Comment (asifdxtreme) [reply csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go]: Fixed

File: csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go (hunk 1 above)
Comment (kumarashit) [reply csi-plug-n-play/sidecars/soda-proxy/pkg/controller/profile/profile.go:17]: Ok

**Comment by asifdxtreme** [comment]
@kumarashit Fixed all the review comments

**Review by kumarashit — APPROVED** [review]
LGTM

File: csi-plug-n-play/sidecars/soda-proxy/Readme.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+soda-proxy is a simple http server which helps to call the api's of SodaFoundation in a simpler way, it bypasses the authentication and directly helps to connect the osdsapi-server using the client provided by sodafoundation/api.
+
+### Build
+Please follow the below steps to build and run the proxy
+
+```go
+go get github.com/sodafoundation/nbp
+
+git checkout csipnp_dev
```
Comment (skdwriting) [inline csi-plug-n-play/sidecars/soda-proxy/Readme.md:11]: Once we merge, this step will be wrong? Can we add a comment

File: csi-plug-n-play/sidecars/soda-proxy/Readme.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+Before running soda-proxy you need to export the below variables.
+
+```go
+
+export OPENSDS_ENDPOINT=http://{your_host_ip}:50040
+export OPENSDS_AUTH_STRATEGY=keystone
+export OS_AUTH_URL=http://{your_host_ip}/identity
+export OS_USERNAME=admin
+export OS_PASSWORD=opensds@123
```
Comment (skdwriting) [inline csi-plug-n-play/sidecars/soda-proxy/Readme.md:26]: Can we avoid this plain text or mention the key env var. Also these env vars can be taken from any config or..?

Comment (asifdxtreme) [reply csi-plug-n-play/sidecars/soda-proxy/Readme.md:26]: these are same env variable which are required for setting up osdsctl https://github.com/sodafoundation/api/wiki/SODA-Projects-Cluster-Installation-through-Ansible#how-to-test-soda-projects-cluster

File: csi-plug-n-play/sidecars/soda-proxy/Readme.md (hunk 3 above)
Comment (asifdxtreme) [reply csi-plug-n-play/sidecars/soda-proxy/Readme.md]: @skdwriting I have removed this step, as we will raise the PR to master today itself.

File: csi-plug-n-play/sidecars/soda-proxy/Readme.md (hunk 4 above)
Comment (asifdxtreme) [reply csi-plug-n-play/sidecars/soda-proxy/Readme.md:24]: the same info has been updated in the Readme

File: csi-plug-n-play/sidecars/soda-proxy/Readme.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+```go
+go get github.com/sodafoundation/nbp
+
+cd $GOPATH/src/github.com/sodafoundation/nbp/csi-plug-n-play/sidecars/soda-proxy
+
+go build cmd/proxy.go
+```
+
+Before running soda-proxy you need to export the below variables.(These are the same env variables which are required by osdsctl, for more reference you can see [this](https://github.com/sodafoundation/api/wiki/SODA-Projects-Cluster-Installation-through-Ansible#how-to-test-soda-projects-cluster))
```
Comment (skdwriting) [inline csi-plug-n-play/sidecars/soda-proxy/Readme.md:16]: This wiki page is deprecated and the latest link is https://docs.sodafoundation.io/soda-gettingstarted/installation-using-ansible/
Could you please confirm once with @kumarashit and update

Comment (kumarashit) [reply csi-plug-n-play/sidecars/soda-proxy/Readme.md:16]: Yes @asifdxtreme Please update the link to docs.sodafoundation. The wiki link is deprecated and discouraged to use :)

Comment (asifdxtreme) [reply csi-plug-n-play/sidecars/soda-proxy/Readme.md]: @skdwriting @kumarashit The links have been updated now

**Review by kumarashit — APPROVED** [review]
LGTM

**Review by skdwriting — APPROVED** [review]
LGTM

**Pull request merged**